--- origsrc/libcxx-10.0.1.src/CMakeLists.txt	2023-07-23 20:21:09.866308800 +0800
+++ src/libcxx-10.0.1.src/CMakeLists.txt	2023-07-23 20:23:52.132094100 +0800
@@ -535,7 +535,9 @@
   # Hide all inline function definitions which have not explicitly been marked
   # visible. This prevents new definitions for inline functions from appearing in
   # the dylib when get ODR used by another function.
+  if (NOT CYGWIN)
   target_add_compile_flags_if_supported(${target} PRIVATE -fvisibility-inlines-hidden)
+  endif()
 
   # Our visibility annotations are not quite right for non-Clang compilers,
   # so we end up not exporting all the symbols we should. In the future, we
