--- origsrc/libcxx-14.0.5.src/include/__config	2023-12-15 16:28:20.030380500 +0800
+++ src/libcxx-14.0.5.src/include/__config	2023-12-15 16:35:17.020095600 +0800
@@ -659,7 +659,7 @@
 #      define _LIBCPP_EXPORTED_FROM_ABI
 #    elif defined(_LIBCPP_BUILDING_LIBRARY)
 #      define _LIBCPP_DLL_VIS __declspec(dllexport)
-#      if defined(__MINGW32__)
+#      if defined(__MINGW32__) || defined(__CYGWIN__)
 #        define _LIBCPP_EXTERN_TEMPLATE_TYPE_VIS _LIBCPP_DLL_VIS
 #        define _LIBCPP_CLASS_TEMPLATE_INSTANTIATION_VIS
 #      else
