--- origsrc/compiler-rt-8.0.1.src/cmake/config-ix.cmake	2019-01-25 17:55:41.000000000 -0500
+++ src/compiler-rt-8.0.1.src/cmake/config-ix.cmake	2019-08-27 00:31:56.442445800 -0400
@@ -587,7 +587,7 @@ else()
 endif()
 
 if (PROFILE_SUPPORTED_ARCH AND NOT LLVM_USE_SANITIZER AND
-    OS_NAME MATCHES "Darwin|Linux|FreeBSD|Windows|Android|Fuchsia|SunOS|NetBSD|AIX")
+    OS_NAME MATCHES "Darwin|Linux|FreeBSD|Windows|CYGWIN|Android|Fuchsia|SunOS|NetBSD|AIX")
   set(COMPILER_RT_HAS_PROFILE TRUE)
 else()
   set(COMPILER_RT_HAS_PROFILE FALSE)
--- origsrc/compiler-rt-8.0.1.src/lib/profile/CMakeLists.txt	2018-07-24 23:01:35.000000000 -0400
+++ src/compiler-rt-8.0.1.src/lib/profile/CMakeLists.txt	2019-08-27 00:32:41.913280800 -0400
@@ -77,7 +77,7 @@ if(WIN32)
     list(APPEND PROFILE_SOURCES WindowsMMap.c)
 endif()
 
-if(FUCHSIA OR UNIX)
+if(FUCHSIA OR UNIX AND NOT CYGWIN)
  set(EXTRA_FLAGS
      -fPIC
      -Wno-pedantic)
